File: src/logic/codefresh.js
import { parse } from '@std/yaml';

export function getCodefreshCredentials() {
    const envToken = Deno.env.get('CF_API_KEY');
    const envUrl = Deno.env.get('CF_BASE_URL');

    if (envToken && envUrl) {
        return {
            headers: { Authorization: envToken },
            baseUrl: `${envUrl}/api`,
        };
    }

    const configPath = Deno.build.os === 'windows'
        ? `${Deno.env.get('USERPROFILE')}/.cfconfig`
        : `${Deno.env.get('HOME')}/.cfconfig`;

    const configFileContent = Deno.readTextFileSync(configPath);
    const config = parse(configFileContent);
    const currentContext = config.contexts[config['current-context']];

    if (!currentContext) {
        throw new Error('Current context not found in Codefresh config.');
    }

    return {
        headers: { Authorization: currentContext.token },
        baseUrl: `${currentContext.url}/api`,
    };
}

export async function getAccountRuntimes(cfCreds) {
    const response = await fetch(`${cfCreds.baseUrl}/runtime-environments`, {
        method: 'GET',
        headers: cfCreds.headers,
    });
    const runtimes = await response.json();
    return runtimes;
}

export async function getRuntimeSpec(cfCreds, runtime) {
    const response = await fetch(`${cfCreds.baseUrl}/runtime-environments/${encodeURIComponent(runtime)}`, {
        method: 'GET',
        headers: cfCreds.headers,
    });
    const runtimeSpec = await response.json();
    return runtimeSpec;
}

export async function getAllAccounts(cfCreds) {
    const response = await fetch(`${cfCreds.baseUrl}/admin/accounts`, {
        method: 'GET',
        headers: cfCreds.headers,
    });
    const accounts = await response.json();
    return accounts;
}

export async function getAllRuntimes(cfCreds) {
    const response = await fetch(`${cfCreds.baseUrl}/admin/runtime-environments`, {
        method: 'GET',
        headers: cfCreds.headers,
    });
    const onPremRuntimes = await response.json();
    return onPremRuntimes;
}

export async function getTotalUsers(cfCreds) {
    const response = await fetch(`${cfCreds.baseUrl}/admin/user?limit=1&page=1`, {
        method: 'GET',
        headers: cfCreds.headers,
    });
    const users = await response.json();
    return users.total;
}

export async function getSystemFeatureFlags(cfCreds) {
    const response = await fetch(`${cfCreds.baseUrl}/admin/features`, {
        method: 'GET',
        headers: cfCreds.headers,
    });
    const onPremSystemFF = await response.json();
    return onPremSystemFF;
}

File: src/cf_support/logic/codefresh.py
from dotenv import load_dotenv
import requests
import yaml
import sys
import os


load_dotenv()


# TODO: Add this when we break out the main function
# required_env_vars: list[str] = ["CF_API_KEY", "CF_URL"]

# for var in required_env_vars:
#     if not var:
#         sys.exit(f"Environment variable {var} is not defined")



def get_runtime_spec(cf_creds, runtime):
    response = requests.get(
        f"{cf_creds['base_url']}/runtime-environments/{runtime}",
        headers=cf_creds["headers"],
    )
    return response.json()


def get_all_accounts(cf_creds):
    response = requests.get(
        f"{cf_creds['base_url']}/admin/accounts",
        headers=cf_creds["headers"],
    )
    return response.json()


def get_all_runtimes(cf_creds):
    response = requests.get(
        f"{cf_creds['base_url']}/admin/runtime-environments",
        headers=cf_creds["headers"],
    )
    return response.json()


def get_total_users(cf_creds):
    response = requests.get(
        f"{cf_creds['base_url']}/admin/user?limit=1&page=1",
        headers=cf_creds["headers"],
    )
    users = response.json()
    return {"totalUsers": users["total"]}


def get_system_feature_flags(cf_creds):
    response = requests.get(
        f"{cf_creds['base_url']}/admin/features",
        headers=cf_creds["headers"],
    )
    return response.json()
